### PR TITLE
Metamist docker tutorial

### PR DESCRIPTION
**Pull request by michael-harper**
This tutorial is to familiarise newcomers to interacting with metamist, building a new Docker image, and submitting a Hail Batch job.

This tutorial is dependent on finalising the dataset to be used and as such, the metamist queries and submission scripts have placeholder `project` names

### CODE REVIEW

**Comment by illusional** [comment]
@michael-harper, now that the sandbox project has been created, is this tutorial completely ready for review? I'm going to dismiss my review as I get daily notifications. If / when it is ready, can you mark this as "Ready for review" and re-request?

Thanks!

File: onboarding_documentation/technical_documentation/metamist_docker/metamist_docker_tutorial.md
Comment (MattWellie) [inline onboarding_documentation/technical_documentation/metamist_docker/metamist_docker_tutorial.md:290]: A ton of the content here could be separated out into an actual script, which you can reference in part (like with links to specific lines). I think you've probably go this script somewhere for the FQE process. Duplicating that code here makes this a massive document for not much extra benefit.

Personally I'd link out to a fully featured script which is mostly self-documenting, and have a section here just explaining the intent

File: onboarding_documentation/technical_documentation/metamist_docker/metamist_docker_tutorial.md
Comment (MattWellie) [inline onboarding_documentation/technical_documentation/metamist_docker/metamist_docker_tutorial.md:420]: This and the following code block also probably exist in the other script, so they could be removed here

File: onboarding_documentation/technical_documentation/metamist_docker/metamist_docker_tutorial.md
Comment (MattWellie) [inline onboarding_documentation/technical_documentation/metamist_docker/metamist_docker_tutorial.md:532]: As previously, it makes sense to explain what this script would do and link to it, rather than recreating a full script in the documentation

**Review by MattWellie — COMMENTED** [review]
Sorry, I've vomited an absolute ton on here

File: onboarding_documentation/technical_documentation/metamist_docker/metamist_docker_tutorial.md
Comment (MattWellie) [reply onboarding_documentation/technical_documentation/metamist_docker/metamist_docker_tutorial.md:231]: Yeah, that would work. If someone wants to read an example but not do the steps themselves, this is great - but the next person can't do this exact same thing

File: onboarding_documentation/technical_documentation/metamist_docker/metamist_docker_tutorial.md
Comment (amyminiter) [inline onboarding_documentation/technical_documentation/metamist_docker/metamist_docker_tutorial.md:4]: It would be nice to specify what 'first hand experience' will entail. What will the user be able to do at the end of the tutorial?  This would help the user select this tutorial if they were looking through documentation.  

How is the tutorial structured and what is the significance of that structure?

File: onboarding_documentation/technical_documentation/metamist_docker/metamist_docker_tutorial.md
Comment (amyminiter) [inline onboarding_documentation/technical_documentation/metamist_docker/metamist_docker_tutorial.md:16]: Suggest defining metamist clearly as CPG's metadata db and providing a link to the documentation. 

What is 'everything' when we are "Keeping track of everything"?

File: onboarding_documentation/technical_documentation/metamist_docker/metamist_docker_tutorial.md
Comment (amyminiter) [inline onboarding_documentation/technical_documentation/metamist_docker/metamist_docker_tutorial.md:15]: What will the user achieve in this section?

File: onboarding_documentation/technical_documentation/metamist_docker/metamist_docker_tutorial.md
Comment (amyminiter) [inline onboarding_documentation/technical_documentation/metamist_docker/metamist_docker_tutorial.md:220]: What will the user achieve in this section?

File: onboarding_documentation/technical_documentation/metamist_docker/metamist_docker_tutorial.md
Comment (amyminiter) [inline onboarding_documentation/technical_documentation/metamist_docker/metamist_docker_tutorial.md:261]: What will the user achieve in this section?

File: onboarding_documentation/technical_documentation/metamist_docker/metamist_docker_tutorial.md
Comment (amyminiter) [inline onboarding_documentation/technical_documentation/metamist_docker/metamist_docker_tutorial.md:726]: What will the user achieve in this section?

File: onboarding_documentation/technical_documentation/metamist_docker/metamist_docker_tutorial.md
Comment (illusional) [inline onboarding_documentation/technical_documentation/metamist_docker/metamist_docker_tutorial.md:260]: I think some ordered steps here might be clearer:

```suggestion
If you wanted to submit this image to be used by the wider CPG:

1. Clone the `images` repo, and create a branch for this image,
2. Create the Dockerfile, you could test the build locally by running a `docker build` (see below),
3. You can issue a test build in the images repository, pushing your image to a container registry that a test account can access,
4. On PR approval and merge, you can deploy your image for use in production workflows.
```

File: onboarding_documentation/technical_documentation/metamist_docker/metamist_docker_tutorial.md
Comment (illusional) [inline onboarding_documentation/technical_documentation/metamist_docker/metamist_docker_tutorial.md:268]: Improving the accuracy of this messaging. They are not technically identical between chipsets, but should be functionally equivalent.

``````suggestion
```bash
docker build -t fastqe_image:0.3.5 . # <-- don't forget the trailing '.'

# If using an M series Macbook, you'll need to specify `--platform=linux/amd64`
docker build \
    -t fastqe_image:0.3.5 \
    --platform=linux/amd64 \
    .
```

Note: We use x86 processors for production workloads (default on GCP), so this flag won't be necessary. Docker files are relatively platform agnostic, meaning that if a build is successful on both your local machine and the GitHub server, the resulting images _should_ be _functionally_ identical. This is beneficial for testing purposes, but it's not a prerequisite for creating images within our infrastructure.
``````

File: onboarding_documentation/technical_documentation/metamist_docker/metamist_docker_tutorial.md
Comment (illusional) [inline onboarding_documentation/technical_documentation/metamist_docker/metamist_docker_tutorial.md:757]: Put these ones on separate lines to remain consistent

File: onboarding_documentation/technical_documentation/metamist_docker/team_docs_metamist_tutorial.py
Comment (illusional) [inline onboarding_documentation/technical_documentation/metamist_docker/team_docs_metamist_tutorial.py:49]:
```suggestion
            sg_assay_map[sg_id] = [read['location'] for read in reads]
```

File: onboarding_documentation/technical_documentation/metamist_docker/team_docs_metamist_tutorial.py
Comment (illusional) [inline onboarding_documentation/technical_documentation/metamist_docker/team_docs_metamist_tutorial.py:1]: I've left comments here instead of the tutorial, so ensure any changes here get reflected above.

File: onboarding_documentation/technical_documentation/metamist_docker/metamist_docker_tutorial.md
Comment (illusional) [inline onboarding_documentation/technical_documentation/metamist_docker/metamist_docker_tutorial.md:276]: Are we intending for users to run this, if so we should add instructions on branching, and committing changes so they can follow along.

**Review by illusional — CHANGES_REQUESTED** [review]
This is very comprehensive @michael-harper, awesome work! I've dotted a few comments, and rewritten some paragraphs where I think greater clarity is beneficial.

I've made suggestions to the script, but not to the guide :) 

I like the ordering of this, though I wonder if it's worth separating this into 3 parts to avoid this single file being actually quite huge for new beginners! This might allow clearer intermediate goals and wins, and might be a useful reference for people coming back to revisit specific sections.

1. introduction to tutorial + metamist
1. writing the image
1. writing the workflow

For the most part the language used here is all good, but there are sections where I think the phrasing could be tightened up, or elements that could be phrased as dotpoints which in some ways forces a distilled points. Happy to chat more about this too!

File: onboarding_documentation/technical_documentation/metamist_docker/team_docs_metamist_tutorial.py
Comment (michael-harper) [reply onboarding_documentation/technical_documentation/metamist_docker/team_docs_metamist_tutorial.py]: Something like this?
```
config = get_config()
project = config["workflow"]["dataset"]
if config["workflow"]["access_level"] == "test" and "test" not in project:
        project = f"{project}-test"
out_web_path = output_path(f"{sgid}-{last_name}.html", "web")
out_display_path = output_path(f"{sgid}-{last_name}.html", "web_url")
```

File: onboarding_documentation/technical_documentation/metamist_docker/metamist_docker_tutorial.md
Comment (michael-harper) [reply onboarding_documentation/technical_documentation/metamist_docker/metamist_docker_tutorial.md]: Have changed accordingly 🙂

File: onboarding_documentation/technical_documentation/metamist_docker/metamist_docker_tutorial.md
Comment (michael-harper) [reply onboarding_documentation/technical_documentation/metamist_docker/metamist_docker_tutorial.md]: Changed!

**Review by amyminiter — COMMENTED** [review]
Good work, @michael-harper. 

I am submitting a partial review after the Docker section. I am not sure how to complete the section of the tutorial and I'll need a few changes before I can complete this and verify the output.

Overall, this section assumes too much previous knowledge of Docker and it needs to be more prescriptive and hand-holdy for the user. 

Please let me know when you've had time to integrate some changes and I'll re-review and have another go. Ta!

File: onboarding_documentation/technical_documentation/metamist_docker/team_docs_metamist_tutorial.py
Comment (MattWellie) [reply onboarding_documentation/technical_documentation/metamist_docker/team_docs_metamist_tutorial.py]: This is the output_path method: [here](https://github.com/populationgenomics/cpg-utils/blob/324fcc2b118f7c125383e7856e1e6e539b922d33/cpg_utils/hail_batch.py#L452)

It handles the dataset, bucket, output prefix, and test/main, with the dataset and access_level referencing the same config entry you're using here. You'd simplify this to
```
html_name = f'{sgid}-{last_name}.html'
out_web_path = output_path(html_name, 'web')
out_display_path = output_path(html_name, "web_url')
```

**Comment by MattWellie** [comment]
I'm going to remove my review request here - there's a lot of chat, but I think perfect is becoming the enemy of the good. I like the intention of this as a readme and a process, I think it's in a pretty good state, and we won't know what about it needs polishing until someone takes it and runs with it. This can still serve as a living document for future edits, so acceptance of the current state doesn't have to be _final_.

I also think there are some distractions around Docker etc. - this is not intended to be an instruction document on how to make docker images, and people not familiar with Docker probably shouldn't be designing images for use in pipelines - thats a separate skillset, and lack of detail there shouldn't hold up this end-to-end guide. A separate document on Docker image design/container usage might be a good idea, but it's far less CPG-specific, so we should defer to public resources like Docker's own [get started guide](https://docs.docker.com/get-started/) for the most part, and pad that out with CPG-specific details where required.

File: onboarding_documentation/technical_documentation/metamist_docker/metamist_docker_tutorial.md
Comment (michael-harper) [reply onboarding_documentation/technical_documentation/metamist_docker/metamist_docker_tutorial.md:301]: Steps 3 and 4 are steps to deploy the image as a usable image in the CPG infrastructure - beyond the scope of this tutorial but outlining the next steps to round out the purpose of building and using a Docker image.

**Review by MattWellie — APPROVED** [review]
👍

File: onboarding_documentation/technical_documentation/metamist_docker/team_docs_metamist_tutorial.py
Comment (michael-harper) [reply onboarding_documentation/technical_documentation/metamist_docker/team_docs_metamist_tutorial.py:57]: Spoken to Matt about this and apparently he has run in to issues with imports being outside functions before. I don't know if there's any official documentation that explains this.

**Pull request merged**
